data/holor.lean

```diff
@@ -0,0 +1,339 @@
+
+lemma mul_right_distrib [distrib α] (x : holor α ds₁) (y : holor α ds₁) (z : holor α ds₂) :
+  (x + y) ⊗ z = x ⊗ z + y ⊗ z :=
+funext (λt, right_distrib (x (holor_index.take t)) (y (holor_index.take t)) (z (holor_index.drop t)))
+
+@[simp] lemma zero_mul {α : Type} [ring α] (x : holor α ds₂) :
+  (0 : holor α ds₁) ⊗ x = 0 :=
+funext (λ t, zero_mul (x (holor_index.drop t)))
+
+@[simp] lemma mul_zero {α : Type} [ring α] (x : holor α ds₁) :
+  x ⊗ (0 :holor α ds₂) = 0 :=
+funext (λ t, mul_zero (x (holor_index.take t)))
+
+lemma mul_scalar_mul [monoid α] (x : holor α []) (y : holor α ds) :
+  x ⊗ y = x ⟨[], forall₂.nil⟩ • y :=
+by simp [mul, has_scalar.smul, holor_index.take, holor_index.drop]
+
+/- holor slices -/
+
+/-- A slice is a subholor consisting of all entries with initial index i. -/
+def slice (x : holor α (d :: ds)) (i : ℕ) (h : i < d) : holor α ds :=
+  (λ is : holor_index ds, x ⟨ i :: is.1, forall₂.cons h is.2⟩)
+
+def unit_vec [monoid α] [add_monoid α] (d : ℕ) (j : ℕ) : holor α [d] :=
+  λ ti, if ti.1 = [j] then 1 else 0
+
+lemma holor_index_cons_decomp (p: holor_index (d :: ds) → Prop):
+  Π (t : holor_index (d :: ds)),
+    (∀ i is, Π h : t.1 = i :: is, p ⟨ i :: is, begin rw [←h], exact t.2 end ⟩ ) → p t
+| ⟨[], hforall₂⟩        hp := absurd (forall₂_nil_left_iff.1 hforall₂) (cons_ne_nil d ds)
+| ⟨(i :: is), hforall₂⟩ hp := hp i is rfl
+
+/-- Two holors are equal if all their slices are equal. -/
+lemma slice_eq (x : holor α (d :: ds)) (y : holor α (d :: ds))
+  (h : slice x = slice y) : x = y :=
+funext (λ t : holor_index (d :: ds), holor_index_cons_decomp (λ t, x t = y t) t (λ i is hiis,
+have hiisdds: forall₂ (<) (i :: is) (d :: ds), begin rw [←hiis], exact t.2 end,
+have hid:     i<d,                             from (forall₂_cons.1 hiisdds).1,
+have hisds:   forall₂ (<) is ds,               from (forall₂_cons.1 hiisdds).2,
+calc
+  x ⟨i :: is, _⟩ = slice x i hid ⟨is, hisds⟩  : congr_arg (λ t, x t) (subtype.eq rfl)
+              ... = slice y i hid ⟨is, hisds⟩  : by rw h
+              ... = y ⟨i :: is, _⟩             : congr_arg (λ t, y t) (subtype.eq rfl)
+))
+
+lemma slice_unit_vec_mul [ring α] {i : ℕ} {j : ℕ}
+  (hid : i < d) (x : holor α ds):
+  slice (unit_vec d j ⊗ x) i hid = if i=j then x else 0 :=
+  if h : i=j then
+    funext (λ t : holor_index ds,
+      by simp [slice, mul, holor_index.take, unit_vec, holor_index.drop, h])
+  else
+    funext (λ t : holor_index ds,
+      by simp [slice, mul, holor_index.take, unit_vec, holor_index.drop, h]; refl)
+
+lemma slice_add [has_add α] (i : ℕ) (hid : i < d)  (x : holor α (d :: ds)) (y : holor α (d :: ds)):
+  slice x i hid + slice y i hid = slice (x + y) i hid := funext (λ t, by simp [slice,(+)])
+
+lemma slice_zero [has_zero α] (i : ℕ) (hid : i < d)  :
+  slice (0 : holor α (d :: ds)) i hid = 0 := funext (λ t, by simp [slice]; refl)
+
+local attribute [instance] classical.prop_decidable
```

Comment:
this is a bad idea. `slice_sum` will now use classical decidability. But if we ever use it on a type `β` which uses another `decidable_eq` then it will hard to apply (as the decidable parameter will not fit).

It is better to assume `decidable_eq β` in `slice_sum`. If you need the law of excluded middle in a proof itself use `by letI := classical.dec_eq β; ...`

Reply:
I tried to do the same in the definition of `cprank`, but it seems that I can't avoid using tactics in the definition then. Is this ok?
```
noncomputable def cprank [ring α] (x : holor α ds) : nat :=
begin 
  letI := classical.dec_pred (λ (n : ℕ), cprank_max n x),
  exact @nat.find (λ n, cprank_max n x) _ ⟨ds.prod, cprank_max_upper_bound x⟩,
end

Reply:
You can avoid it, but directly providing the `classical.dec_prec ...` proof to `nat.find`.

Reply:
Oh, that's easier than I thought :D